### ISSUE TITLE
Forecast: Research suitable forecast models

### ISSUE THREAD

**Issue by DonWillems** [openremote/openremote#1599]
Find suitable ML model, see use cases #1309

**Comment by Hackerberg43** [comment]
I've recently attended a presentation from Saxion/UTwente where they researched this topic for electricity forecasting.
Here is a link to the paper:  https://essay.utwente.nl/100847/1/TRIVEDI_BA_EEMCS.pdf
One side note they made, is that they didn't consider deep learning algorithms (like xgboost) and focused on statistical models with low computational cost. Their winning contender was Prophet.


Might help with your research :)

**Comment by dominiquekleeven** [comment]
**NOTE**: I am still making some edits to this comment, in particular to XGBoost section since I am still working on the performance metrics of the forecast that takes into account actual historial weather data and future weather predictions. The conclusion/recommendation remains the same.

# Model for IoT Focused Forecasting

I've gone through various papers on forecasting in energy, predictive maintenance, and power usage trends. Also looked at blogs/resources that explore this topic and the following models were most present/discussed as suitable ML models for time series based forecasting:

- Prophet
- XGBoost
- LTSM

Information and a summary of these models can be found below, for `Prophet` and `XGBoost` I've also setup Jupyter Notebooks to explore the models and try them out on datasets that we have access to and determine their performance metrics.

---

## **Prophet (Facebook/Meta)**
**Type:** Time Series Forecasting

#### **Pros:**
- Handles strong seasonal trends well.
- Works with missing data and outliers.
- Gives clear insights into trend, seasonality, and holiday effects.
- Fast and scalable, making it suitable for IoT applications.
- Easy to use and can be tuned for specific holiday effects.

#### **Cons:**
- Doesn't perform well with 'unpredictable' data
- Not as flexible as more complex models like LSTM or XGBoost when dealing with non-seasonal trends.
- Can be less accurate for very short-term forecasting (minutes to hours).
- Not flexible, e.g. when providing regressors to train on, then these must also be provided as part of the forecast window data. Meaning you'd need future data for these regressors.
- Python/R only from what I can find (no bindings for Java)

#### **Use Cases:**
- Forecasting electricity demand based on seasonal trends.
- Predicting solar/wind power generation.
- Long-term energy consumption analysis.

### Testing / Exploration Results
**Note:** I've setup a Jupyter Notebook for this using data from a production environment + historical weather data from openmeteo a 14 day weather forecast)

#### Results without any regressors: (Generally performs bad)
(A regressor is an external variable that 'might' influence the target variable)
- `RMSE = 0.042493`Average prediction error of about 4.2 cents per kWh. In real terms: if actual price is €0.15, model might predict €0.19 or €0.11
- `R² = -0.365033` A negative value means model performs worse than just using the mean value


#### Results with regressors (cloud_cover, wind_speed, shortware radiation, rush_hour, is_day)
(A regressor is an external variable that 'might' influence the target variable)
- `RMSE = 0.031150`Average prediction error of about 3.1 cents per kWh. In real terms: if actual price is €0.15, model might predict €0.18 or €0.12
- `R² = 0.266467` A positive value means model better than using average/mean values.

Overall performance will increase as more regressors are added and more historical data is available. Because the Prophet model is 'plug and play' its performance is quite decent.

---

## **XGBoost**
**Type:** Supervised ML (Gradient Boosting)

#### **Pros:**
- High accuracy and efficiency, very fast to train.
- Handles missing data/lack of data quite well
- Can capture non-linear relationships and interactions between features/variables.
- Well-suited for tabular data (sensor data, weather data, etc.).
- Has libraries available for Java (Allowing for a direct integration path)

#### **Cons:**
- Might require careful feature engineering to achieve optimal forecasts. (no plug and play)
- Can require a lot of compute for very large datasets. (basically 500k+ records/rows datasets)
- Performance depends heavily on parameter tuning and cross-validation.
- Harder to understand than simpler models like Prophet.

#### **Use Cases:**
- Short-term power consumption prediction using sensor data.
- Fault detection and predictive maintenance
- Estimating energy demand based on external factors (weather, occupancy, etc.).

### Testing / Exploration Results (Energy Tariff Forecast)
**Note:** I've setup a Jupyter Notebook for this using data from a production environment + historical weather data from openmeteo a 14 day weather forecast)

#### Results with basic features: (Generally performs OK, features: is_peak, is_night, wind_speed)
- `RMSE: 0.0337` Average prediction error of about 3.3 cents. In real terms: if actual price is €0.15, model might predict €0.18 or €0.12
- `R² Score: 0.2273` A positive value means the model performs better than just using the mean/average value.

#### Results with more complex features (historical weather data, future weather data, rolling windows)
<< Work in progress >> 
Work in progress, still being tweaked to properly use weather forecast data from openmeteo

---

## **LSTM (Long Short-Term Memory)**
**Type:** Deep Learning (Recurrent Neural Network)

I was quick to determine that LTSM and similar models aren't really suitable for smaller data sets or dealing with real-time changes. This is mainly because of the following reasons:
- It requires  large datasets to perform well, which may not always be available.
- Computationally expensive to train, requiring significant hardware resources (e.g., GPUs).
- Training time can be long, making it less suitable for real-time predictions.

Therefore I've chosen to not take a deeper look at the LTSM model
---

# Conclusion

- **Prophet** is great for seasonal-based forecasting, especially in scenarios with clear seasonal patterns. Is able to take into account external variables such as the weather forecast.
- **XGBoost** is ideal for short-term forecasts, especially when working with sensor data and various external factors, whilst also being fast to train. More flexible when dealing with a lack of specific data.
- **LSTM** should only be considered for very large datasets and specific usecases where enough data and compute is available.

Also important to note both Prophet and XGBoost were quite fast to train and didn't use a lot of compute/memory. Al though the dataset was quite limited (3 months). Less than 10 seconds usually, depending on the amount of regressors/features.

## My recommendation
- **Prophet** for a more plug-and-play system, doesn't require a lot of technical know how from both a developers and users perspective, surface level knowledge should be enough.
- **XGBoost** for a better performance in terms of forecasting, customizability and flexibility. Important to note its setup is complex for both the developer and also the user when it comes to customizing and setting up their forecast model. Requires more in-depth knowledge about the data, domain and machine learning.

**Ideally the option between XGBoost/Prophet is discussed within the team taking into account their considerations (complexity for the user in particular)**

#### Implementation considerations:
- Leveraging existing platforms (AWS Sagemaker) that allow us to train models on demand and receive forecasts from them: https://aws.amazon.com/sagemaker/
- An optional deployable service that integrates into OpenRemote
- Internal service part of the backend, would require a ML Model that can be used with java, XGBoost has a library for this, but Prophet does not.


Sources used:
- https://essay.utwente.nl/100847/1/TRIVEDI_BA_EEMCS.pdf 
- https://www.geeksforgeeks.org/time-series-analysis-using-facebook-prophet/
- https://arxiv.org/pdf/2306.09129
- https://facebook.github.io/prophet/
- https://www.kaggle.com/code/robikscube/time-series-forecasting-with-prophet-yt (notebook with examples)
- https://ieeexplore.ieee.org/document/10673558
- https://www.mdpi.com/2227-7390/11/5/1068
- https://xgboost.readthedocs.io/en/stable/
- https://arxiv.org/pdf/2107.06174

**Comment by dominiquekleeven** [comment]
We discussed the above briefly during the sprint meeting. In regards to implementation it makes sense to go towards a seperate service (e.g. microservice) that exposes an API to interact with the forecasting models. This also allows us to use Python which has the majority in terms of library support when it comes to anything ML/AI/Data.
